### PR TITLE
Implement SMB3 Signing (AesCmac), 3.0.x encryption (AES/CCM) and 3.1.1 encryption (AES/GCM)

### PR DESCRIPTION
**Pull request by yin19941005**
Implement SMB3 Signing (AesCmac), 3.0.x encryption (AES/CCM) and 3.1.1 encryption (AES/GCM)

Fix Cancel Request didn't send sessionId. Added NegotitateConext classes for 3.1.1 and applied them on negotiate request and response. Added transformHeader for SMB3 encryption and its packet factory for decrypt. Added PacketEncryptor for packet encryption. Added AEADBlockCipher for AES/CCM/NoPadding and AES/GCM/NoPadding, and this may integrate with the old cipher interface in the future. Handle the session and treeConnect flags on response for requiring encryption, so that, the library is able using plainText to access other share but only required encrypt share will be encrypted. Added AesCmac support in PacketSignatory for SMB3 signing. CryptographicKeysGenerator is added for key generate but not yet passed the security provider to it and always using BC classes directly. DigestUtil is added for 3.1.1 preAuthHashValue and the session authenticate process had always changed in order to calculate the preAuthHashValue for 3.1.1.

Hi @hierynomus,
I think this pull request should not directly merge to master right now. It should merge to a feature branch so that more testing and improvement could be done. However, it doesn't seems github provide me any option for doing this. What do you think about it?

Demo config for using dialect 3.0.2
```
SmbConfig config = SmbConfig.builder()
            .withDialects(SMB2Dialect.SMB_3_0_2)
            .withSecurityProvider(BCSecurityProvider())
            .withCapabilities(SMB2GlobalCapability.SMB2_GLOBAL_CAP_DFS, SMB2GlobalCapability.SMB2_GLOBAL_CAP_ENCRYPTION)
            .withEncryptData(true)
            .build()
```

Demo config for using dialect 3.1.1, adding BouncyCastleProvider is recommended on 3.1.1 as JceSecurityProvider had optimization for GCM.
```
Provider[] providerList = java.security.Security.getProviders();
BouncyCastleProvider bcSecurityProvider = new BouncyCastleProvider();
java.security.Security.insertProviderAt(bcSecurityProvider, providerList.length);

SmbConfig config = SmbConfig.builder()
            .withDialects(SMB2Dialect.SMB_3_1_1)
            .withCapabilities(SMB2GlobalCapability.SMB2_GLOBAL_CAP_DFS, SMB2GlobalCapability.SMB2_GLOBAL_CAP_ENCRYPTION)
            .withEncryptData(true)
            .build()
```

Related to #151 and it should able to solve the issue. Related to #287.

### CODE REVIEW

**Comment by gfenyvesi** [comment]
Hi,
It seems that this branch cannot be accessed, and I would like to use this library to connect to Azure storage.
Can I ask you to make it available somehow, or can you send me the source code or the jar?
Thanks,
Gabor

**Comment by yin19941005** [comment]
Hi @gfenyvesi,

I am not sure what do you mean "cannot be accessed". This merge request is not merged yet and still under development. If you want to run some test on it right now, you may get it from my fork repo: https://github.com/yin19941005/smbj/tree/implement-smb3-signing-encryption.

Any help/test/opinion are are appreciated!

**Comment by gfenyvesi** [comment]
Hi,
I could not find your repo until you shared the link, thank you.
I built the jar and tested this branch and so far so good: I tried getting file information, reading the contents of files and deleting files on Azure and all worked.
But could you let me know if you are aware of any issues? Have you done tests too? Do you think we can use it in production?
Thanks again,
Gabor

**Comment by yin19941005** [comment]
Nice to hear it had been tested and works on Azure!

I tested some basic file operation (read/write, rename, delete, etc.) on Windows 2012 R2 and 2016 as well and it works fine.

There are numbers of concern:

- The dialect compatibility for 3.0.x and 3.1.1
Not sure excepting the signing and encryption, anymore required feature to make 3.0.x and 3.1.1 works. So far based on the testing result, it seems only signing and encryption is the basic requirement to make them works.

- The dialect backward compatibility
Did I mishandle some case and sending some 3.x.x only feature/packet when 2.x.x dialect is negotiated? Did I break anything on 2.x.x dialect? So far tested, didn't found any yet.

- nonce reuse issue (after maximum reached)
I just found this issue recently. It basically wouldn't happen as the nonce length is 11 bytes for CCM and 12 bytes for GCM while the message ID only having 8 bytes (i.e. the message ID will use up before the nonce use up in most of the time). Not sure hot to handle this yet, but reconnect is the simplest way (base on current implementation) to resolve this issue.

- Code cleanup
I believe there are lots of works here (including the code formatting).

**Comment by ineffyble** [comment]
It would be amazing and very helpful if this PR was merged. Is there anything others can do to get this over the line?

**Comment by yin19941005** [comment]
Hi,

Let me explain a bit more to let others more easily to pick up this PR.

#### Security concept related to SMB2 on this PR

- **nonce** (in 2.2.41 SMB2 TRANSFORM_HEADER), required to unique within a session, i.e. we should never use random nonce in our implementation. You may check [this paper](https://eprint.iacr.org/2016/475.pdf) out for more details.

- **CCM** and **GCM** mode are authenticated encryption algorithm which designed to provide both data integrity (authenticity ) and confidentiality. They both have authentication tag and already guarantee the data integrity, i.e. the **signing is not required** if the data is encrypted with these mode, and SMB2 document had clear state we must not sign the packet and must skip the signature verification if the packet is encrypted/decrypted. 

#### Changes in Code
For SMB3 Signing:
1. Added `AesCmac` algorithm for SMB3 Signing in `com.hierynomus.security.bc.BCMac` class
2. Implement `com.hierynomus.security.CryptographicKeysGenerator` class for Signing key generation.
3. Change `com.hierynomus.smbj.session.PacketSignatory` to generation signing (by CryptographicKeysGenerator) and using correct sign algorithm for different dialect.

For SMB 3.0.x encryption:
1. Add `com.hierynomus.mssmb2.Smb2EncryptionCipher` enum class for CCM and GCM type. Also get the algorithm name to pass to the security provider in this enum class.
2. Add `isFromDecrypt` Boolean for both `com.hierynomus.mssmb2.SMB2PacketData` and `com.hierynomus.mssmb2.SMB2Packet`. If this value is true, the signing verification will be skipped.
3. Change `verify` method in `com.hierynomus.smbj.session.PacketSignatory`, always return true if packet `isFromDecrypt` is true.
4. Change `verifyPacketSignature` method in `com.hierynomus.smbj.connection.Connection`, don't throw exception when signed flag is not set and signing required is true when packet's `isFromDecrypt` is true.
5. Add `isRequireEncrypt` Boolean for `com.hierynomus.mssmb2.SMB2Packet` to indicate the packet must be encrypted before sending to the server.
6. Add `com.hierynomus.mssmb2.SMB2TransformHeaderFunctions` class, helper functions for creating  `SMB2 TRANSFORM_HEADER`.
7. Implement `com.hierynomus.smbj.session.PacketEncryptor` to encrypt the packet and wrapped it as `SMB2 TRANSFORM_HEADER` when write. **This is the class that the encryption occur**.
8. Change `com.hierynomus.mssmb2.messages.SMB2CancelRequest` to ensure it able to get correct session for encryption.
9. Change `com.hierynomus.smbj.SmbConfig`, to let user able to set client capabilities and decided to encrypt data or not if the server is not requiring to encrypt the data. Also check the input parameter is valid or not (for encryption) when build the config.
10. Add `com.hierynomus.mssmb2.SMB2ShareFlags` enum to handle the share flags on the treeConnect response. The server may set `SMB2_SHAREFLAG_ENCRYPT_DATA` and forced client to encrypt data on this share.
11. Add `encryptData` Boolean in `com.hierynomus.smbj.share.TreeConnect` to handle the `SMB2_SHAREFLAG_ENCRYPT_DATA` flag correctly.
12. Change `com.hierynomus.smbj.share.TreeConnect`. Add `sendBySession` method and use it to send the close request. This is ensuring the `TREE_DISCONNECT Request` is encrypted as well.
13. Change `com.hierynomus.smbj.share.Share`, ensure the `isRequireEncrypt` Boolean is correctly set for sending packet.
14. Add `com.hierynomus.security.DecryptPacketInfo` data class to hold the decryption info, i.e. decryption key and decryption algorithm.
15. Add `com.hierynomus.mssmb2.SMB2SafeIgnorePacketData` in order to handle the cases which the SMB2 document state the client need to discard response packet.
16. Implement `com.hierynomus.mssmb2.SMB3EncryptedPacketFactory` to parse the `SMB2 TRANSFORM_HEADER` and decrypt the server response. After the decryption, the plaintext packet will pass to the SMB2MessageConverter to handle. `isFromDecrypt` will be set in here as well. **This is the class the decryption occur**.
17. Add `sessionDecryptInfoMap` in `com.hierynomus.smbj.connection.Connection`. This allow us retrieve the correct decryption key for different session.
18. Change the message converter in `com.hierynomus.smbj.connection.Connection` no longer static. This is because different connection may have different security provider for encrypt/decrypt message. Generate the decryption key and put it to `sessionDecryptInfoMap`.
19. Change `com.hierynomus.smbj.connection.ConnectionInfo` to determine the connection is supported encryption or not.
20. Add `packetEncryptor` as private field in `com.hierynomus.smbj.session.Session`. Set the sessionKey for it to generate the encryption key. Handling the encryption logic as well.
21. Add `AEADBlockCipher` (Authenticated Encryption with Associated Data, AEAD) in `com.hierynomus.security` package for CCM (also GCM in 3.1.1). This not necessary separate with the original `Cipher` class, separate them is just for easier implementation.
22. Change `com.hierynomus.mssmb2.messages.SMB2NegotiateRequest` for passing the clientCapabilities in config.
23. Remove all the TODO or exception failing the SMB 3.0.x dialect.

For SMB 3.1.1 encryption:
Basically all the remaining stuffs, e.g. SMB2NegotiateContext, pre-auth Hash, Smb2HashAlgorithm, etc. Will complete the list later on.

**Comment by hierynomus** [comment]
@yin19941005 I'll merge this onto the `smb3-dev` branch before merging it into master. I'll also update that branch with the current master.

We can then close this PR and create a new one to continue the work. For me a few more tests are required to ensure the coverage doesn't go down too much, before we merge this into master and release it.

**Comment by hierynomus** [comment]
BTW: Thanks for all the hard work!

**Comment by SubhasishDey** [comment]
> Hi,
> I could not find your repo until you shared the link, thank you.
> I built the jar and tested this branch and so far so good: I tried getting file information, reading the contents of files and deleting files on Azure and all worked.
> But could you let me know if you are aware of any issues? Have you done tests too? Do you think we can use it in production?
> Thanks again,
> Gabor

Hi Gabor,

             Is it possible for you to share the code snippet, which you used to connect to the Azure Share ?
I tried below one but it doesnot work :

`SmbConfig smbConfig = SmbConfig.builder()
					.withMultiProtocolNegotiate(true)
					.withDfsEnabled(true)
					.withDialects(SMB2Dialect.SMB_3_1_1)
					.withCapabilities(SMB2GlobalCapability.SMB2_GLOBAL_CAP_DFS, SMB2GlobalCapability.SMB2_GLOBAL_CAP_ENCRYPTION)
					.withEncryptData(true)
					.build();`

Thanks and Regards,
Subhasish

**Comment by gfenyvesi** [comment]
Well, I can share only parts of it, because it is integrated into a proprietary code base. Even I am not sure if this works with Azure, because that test was a temporary project and already deleted.

We have different implentations of a „File”, here are relevent parts from the Smb implementation:

    public synchronized DiskShare openDiskShare() throws IOException {
        if(client == null) {
            client = new SMBClient(
                SmbConfig.builder()
                    .withSocketFactory(new ProxySocketFactory(60 * 1000))
                    .withDfsEnabled(true)
                    .build()
            );
        }
        if(connectionForClose == null) {
            connectionForClose = client.connect(hostNameAndShare.getO1());
        }
        if(sessionForClose == null) {
            String domain = null;
            String login = TODO;
            if(login != null && login.length() > 0) {
                int atPos = login.indexOf('@');
                if(atPos >= 0) {
                    domain = login.substring(atPos + 1);
                    login = login.substring(0, atPos);
                } else {
                    int slashPos = login.indexOf('\\');
                    if(slashPos >= 0) {
                        domain = login.substring(0, slashPos);
                        login = login.substring(slashPos + 1);
                    }
                }
            }
            AuthenticationContext ac = new AuthenticationContext(
                login
                , password == null || password.isEmpty() ? new char[0] : password.toCharArray()
                , domain
            );
            sessionForClose = connectionForClose.authenticate(ac);
        }
        if(diskShareForClose == null) {
            diskShareForClose = (DiskShare) sessionForClose.connectShare(hostNameAndShare.getO2());
            if(log.isDebugEnabled()) {
                log.debug("SMB Share has been opened " + diskShareForClose);
            }
        }
        return diskShareForClose;
    }

    private DiskEntry openSmbFile(
        Set<AccessMask> accessMasks
        , Set<SMB2ShareAccess> shareAccesses
        , SMB2CreateDisposition disposition
    ) throws IOException {
        if(smbFileForClose != null) {
            return smbFileForClose;
        }
        DiskShare diskShare = openDiskShare();
        smbFileForClose = diskShare.open(
            url
            , accessMasks
            , null
            , shareAccesses
            , disposition
            , null
        );
        if(log.isDebugEnabled()) {
            log.debug("SMB File has been opened " + smbFileForClose + " for " + disposition);
        }
        return smbFileForClose;
    }


    @Override
    InputStream openStream() throws IOException {
        DiskEntry smbFile = openSmbFile(
            EnumSet.of(AccessMask.GENERIC_READ)
            , EnumSet.of(SMB2ShareAccess.FILE_SHARE_READ)
            , SMB2CreateDisposition.FILE_OPEN
        );
        InputStream ret = ((com.hierynomus.smbj.share.File) smbFile).getInputStream();
        dontClose = true;
        return ret;
    }

    @Override
    OutputStream openOutputStream() throws IOException {
        basicInfoCached = null;
        DiskEntry smbFile;
        try {
            smbFile = openSmbFile(
                EnumSet.of(AccessMask.GENERIC_READ, AccessMask.GENERIC_WRITE)
                , EnumSet.of(SMB2ShareAccess.FILE_SHARE_READ, SMB2ShareAccess.FILE_SHARE_WRITE)
                , SMB2CreateDisposition.FILE_CREATE
            );
        } catch(Exception e) { // NOSONAR
            smbFile = openSmbFile(
                EnumSet.of(AccessMask.GENERIC_READ, AccessMask.GENERIC_WRITE)
                , EnumSet.of(SMB2ShareAccess.FILE_SHARE_READ, SMB2ShareAccess.FILE_SHARE_WRITE)
                , SMB2CreateDisposition.FILE_OVERWRITE
            );
        }
        OutputStream ret = ((com.hierynomus.smbj.share.File) smbFile).getOutputStream();
        dontClose = true;
        return ret;
    }

Regards,
Gabor

**Comment by felixRajkumar** [comment]
Hi, I am reading files from remote servers. Can I build smb3-dev and use it ? . Or are there any major issues in encryption development.
